### ISSUE TITLE
Deprecate Auth?

### ISSUE THREAD

**Issue by jkarni** [haskell-servant/servant#805]
The work we put into authentication was really useful in getting the internals right; and exposing an API made sense while nothing else existed. But now `servant-auth` exists and seems to me like the better solution. Can we deprecate the auth stuff in this repo, and start pointing people to `servant-auth` instead?

**Comment by alpmestan** [comment]
I'm a little bit bothered by the fact that authentication is a key topic, so much that it's covered in the tutorial, but you're suggesting to move it _all_ to non-core packages. I _want_ auth to be documented in the tutorial, whichever package it ends up being implemented in. servant-auth's README and the tutorial hold different standards wrt depth of explanation, obviously, so I do not want for the go to auth solution to be underdocumented. It'd have to be covered in the tutorial, in reasonable depth.

So yeah, it might require a little bit of work but nothing too scary, so yes from me pending the changes that I suggest.

**Comment by alpmestan** [comment]
There's also the issue of maintenance. Would the auth packages move under the haskell-servant org? If not, what kind of maintenance/development/collaboration scheme would be in place?

**Comment by jkarni** [comment]
I agree with the tutorial needing to be updated; I'd be happy to do it.

I can check whether we can move 'servant-auth' to the 'haskell-servant'
organization.

**Comment by alpmestan** [comment]
The motivation for my question is that there are going to be changes that break servant-auth, and since auth is too important to leave broken for a bit, we need to make sure we'll be able to get the necesary changes in and probably have releases happen together.

**Comment by 3noch** [comment]
> The motivation for my question is that there are going to be changes that break servant-auth, and since auth is too important to leave broken for a bit

I agree. However, this cuts both ways. Because it's so important you may want to make it one of the official servant libraries. But also, because its so important, if a new release of servant does not work with servant-auth, you can be very sure that people will get if fixed quickly. Until they do, they'll have to use an older version of servant. That scenario is pretty common in package management and doesn't scare me in the slightest.

**Comment by alpmestan** [comment]
@3noch I'd be in favor of having them as core packages (but just not in the standard ones), but it's not my call =) You make good points. My main request is really to have a detailed and very, very approachable section in the tutorial, like the other sections of it. The rest is just about us trying to make sure both package sets will work well together, hand in hand.

**Comment by 3noch** [comment]
@alpmestan You're right about documentation though. That is crucial and it definitely needs to cover auth. I don't care where the package lives, myself. As long as it's easily discoverable by servant neophytes.

**Comment by phadej** [comment]
From my "release-manager" POV, having `servant-auth` as a core package is :+1:. There is no additional work. I (did/will) check how `servant-auth` compiles during preparation for `servant` releases, as it's good sanity test (among others).

In fact, `servant-auth` was `GHC-8.2.1/servant-0.11` ready at the moment `servant-0.11` was released. It was only pending release of external dependencies. And releasing all at once (i.e. major servant, minor servant-auth updates) is less context switching.

So overall, :+1: to moving `servant-auth` to `haskell-servant` organisation.

**Comment by aaronlevin** [comment]
As the creator of this mess, I'm supportive :+1: :)

I wouldn't worry about breaking changes. We made the `Experimental` module for a reason.

**Comment by roberth** [comment]
I have accidentally used this module in a new project. I have switched to servant-auth because it supports optional authentication. Please remove this module. This is "Bad UX" ;)

**Comment by alpmestan** [comment]
This supports optional authentication too, I think? Also, we need a little bit of a migration guide, as there are dozens of projects out there using the auth module from the core packages.

**Comment by roberth** [comment]
Ah, yes, I though the parameter had to be `WithMetadata a` where `a` is some kind of proof of authentication, but that's only in `servant-auth-cookie`. This is a `servant` issue and the `servant` interface seems very unconstraining indeed.

A migration guide is a good idea. In the meanwhile you could move this module into another package, but that seems like too much effort. I think updating the servant tutorial is a good start.

**Comment by alpmestan** [comment]
Right, I think this would be a reasonable plan to migrate the auth stuffs:
- (if Plow Technologies agrees) move servant-auth-* into core packages or get them under the servant org or keep it where it is but make sure we'll be able to sync up as regularly as needed for releases, breaking changes and all that.
- rewrite the auth bit of the tutorial to explain how to use servant-auth, including any instructions about adding dependencies (in case servant-auth-* is not merged in the core packages) etc
- maybe include a small migration guide that shows how to switch something that uses the .Experimenta.Auth over to servant-auth, at least for another major release or two.
- pull the trigger.

**Comment by dbaynard** [comment]
I'm in the process of setting up a server using servant-auth. I'd be happy to contribute to the tutorial, say.

It would be _very_ helpful to have a (referenced) list of the security features within the docs, not least because it shows those of you who have designed the libraries have considered them.

**Comment by domenkozar** [comment]
Talking to @phadej and @alpmestan, here's a straw-man proposal:

- [ ] bring documentation in servant-auth on-pair with servant standards
- [ ] deprecate Servant.API.Experimental.Auth
- [ ] move `Servant.Auth.Auth` and `Servant.Auth.BasicAuth` to `Servant.API.Auth`
- [ ] Make sure `Servant.API.BasicAuth` has feature parity with the new BasicAuth

Then all of the rest of authentication schemes would stay in servant-auth.

Open questions:
- `servant-auth-client` shouldn't depend on http-client due to ghcjs
